### PR TITLE
[WINDUPRULE-792] Add rule for sun.misc.Unsafe methods which have been removed

### PR DESCRIPTION
**Pull request by jmle**
See [WINDUPRULE-792](https://issues.redhat.com/browse/WINDUPRULE-792)

### CODE REVIEW

File: rules-reviewed/openjdk11/openjdk8/java-removals.windup.xml [hunk 1]
```diff
@@ -105,5 +105,29 @@
...
+            <when>
+                <javaclass references="sun.misc.Unsafe.{method-name}({*})"/>
+            </when>
+            <perform>
+                <hint title="Methods in `sun.misc.Unsafe` have been removed" effort="2" category-id="mandatory">
+                    <message>
+                        The `sun.misc.Unsafe` API is scheduled for removal in the long term, and some of its methods have been removed. In short term, please consider the following:
+
+                        - The `get` and `put` methods have equivalents that take long offsets (OpenJDK 8's version of the removed methods just casts the int offsets and used the long versions).
+                        - `staticFieldOffset(Field)` and `ObjectFieldOffset(Field)` should replace `fieldOffset(Field)` (which the 8 version delegated to already).
```
Comment (carlosthe19916) [inline rules-reviewed/openjdk11/openjdk8/java-removals.windup.xml:118]: I understand this message comes from the Ticket itself. But usually, when there are docs mentioning replacements it follows the pattern: `A should be replaced by B` this way it is easy to understand the change `source => target`. The message "staticFieldOffset(Field) and ObjectFieldOffset(Field) should replace fieldOffset(Field)" follows the oposite pattern: `B should replace A`, the pattern is `target => source`. I would suggest to reorder the text. However, I acknowledge English is not my native language so this is just a suggestion and not a required change.

File: rules-reviewed/openjdk11/openjdk8/java-removals.windup.xml [hunk 2]
```diff
@@ -105,5 +105,29 @@
...
+                        - `tryMonitorEnter(Object)` and `monitorExit(Object)` can be replaced with use of the Java synchronized primitives, as illustrated in the test changes for JDK-8054494 (see links below).
+
+                        In the long term, please consider using Variable Handles (see the link below)
+                    </message>
+                    <link href="https://hg.openjdk.java.net/jdk9/jdk9/jdk/rev/0a0a0986400e#l2.19" title="Test changes for JDK-8054494"/>
+                    <link href="https://openjdk.java.net/jeps/193" title="Variable Handles"/>
+                </hint>
+            </perform>
+            <where param="method-name">
+                <matches pattern="(getInt|putInt|getFloat|putFloat|getDouble|putDouble|getBoolean|putBoolean|getObject|putObject|getByte|putByte|getChar|putChar|getShort|putShort|getLong|putLong|fieldOffset|staticFieldBase|tryMonitorEnter|monitorExit)"/>
```
Comment (carlosthe19916) [inline rules-reviewed/openjdk11/openjdk8/java-removals.windup.xml:129]: The methods removed are `fieldOffset(), monitorEnter(), monitorExit(), tryMonitorEnter()`. I do see the `<matches pattern=""/>` tag of the rule you created covers 3 of the 4 methods mentioned before, the one not included is `monitorEnter()`. Although it is not mentioned in the JIRA Ticket do you think it is worth adding it to the Rule?

**Review by carlosthe19916 — CHANGES_REQUESTED** [review]
Thanks for the PR @jmle . My following comments mainly comes from the content of this blog https://blogs.oracle.com/javamagazine/post/the-unsafe-class-unsafe-at-any-speed

- Shouldn't we have a Hint or Message (I don't know exactly what) to let the user know that he needs to manually enable the Java Module `jdk.unsupported` in order to continue using the package `sun.misc`? Perhaps also adding an example? I'm not sure if this can be out of the scope of this PR so we can also do this in other PR or Ticket. What do you think?
- The `defineClass()` method was also moved to the MethodHandles. Do you think we can add a rule or enhance the current one to let the user know that this particular method has been moved and he needs to rewrite the code and start using `MethodHandles`? As before, this might be also out of the scope of the PR so we can do it here or in other PR.
-

File: rules-reviewed/openjdk11/openjdk8/java-removals.windup.xml (hunk 1 above)
Comment (jmle) [reply rules-reviewed/openjdk11/openjdk8/java-removals.windup.xml:118]: Will do :+1:

File: rules-reviewed/openjdk11/openjdk8/java-removals.windup.xml (hunk 2 above)
Comment (jmle) [reply rules-reviewed/openjdk11/openjdk8/java-removals.windup.xml:129]: I agree, `monitorEnter` should have also been added and wasn't included in the ticket. Will do.

**Comment by jmle** [comment]
> Thanks for the PR @jmle . My following comments mainly comes from the content of this blog https://blogs.oracle.com/javamagazine/post/the-unsafe-class-unsafe-at-any-speed
> 
> * Shouldn't we have a Hint or Message (I don't know exactly what) to let the user know that he needs to manually enable the Java Module `jdk.unsupported` in order to continue using the package `sun.misc`? Perhaps also adding an example? I'm not sure if this can be out of the scope of this PR so we can also do this in other PR or Ticket. What do you think?
> * The `defineClass()` method was also moved to the MethodHandles. Do you think we can add a rule or enhance the current one to let the user know that this particular method has been moved and he needs to rewrite the code and start using `MethodHandles`? As before, this might be also out of the scope of the PR so we can do it here or in other PR.

Hi @carlosthe19916, I mainly followed the ticket in this rule. These are good questions, but I think that they probably should be answered by @gnu-andrew, since he was the one who created the issue and is more knowledgeable about it than me.

**Comment by jmle** [comment]
I have created [https://issues.redhat.com/browse/WINDUPRULE-831](https://issues.redhat.com/browse/WINDUPRULE-831) to potentially deal with @carlosthe19916 suggested enhancements.

**Comment by carlosthe19916** [comment]
> I have created https://issues.redhat.com/browse/WINDUPRULE-831 to potentially deal with @carlosthe19916 suggested enhancements.

Thanks, @jmle for creating the ticket https://issues.redhat.com/browse/WINDUPRULE-831  . I will approve this PR since it does what has been requested by the Ticket. We need to keep an eye on the Ticket you have just created to decide what to do with it. 

@gnu-andrew when you find convenient, may I ask you to give us some feedback about the ticket that has been created please? https://issues.redhat.com/browse/WINDUPRULE-831

**Review by carlosthe19916 — APPROVED** [review]

**Pull request merged**